### PR TITLE
New type StrLimited

### LINKED ISSUE #2104: Simplified length limited string
I often want `constr(max_length=XXX, strip_whitespace=True)`. Could we add a shortcut for this that worked better with mypy?

E.g. something like

```py
class Foobar(BaseModel):
    name: StrLimit[63]
```

This would be equivalent to `constr(max_length=63, strip_whitespace=True)`, e.g a field with a max length of `63`, no min limit, `strip_whitespace` enabled.

The name `StrLimit` is debatable. I think using `ConStr` might be confusing, but perhaps would make more sense?

We could also allow `StrLimit[10, 63]` which would be equivalent to `constr(min_length=10, max_length=63, strip_whitespace=True)`. However this would not be very pythonic in that the first argument meaning would change. Also I find myself wanting this very rarely.

thoughts?

### PR DESCRIPTION
**Pull request by lucastosetto**
## Change Summary

This represents a simplified length limited string and is an alias for the following syntax:

`StrLimited[max_length]` = `constr(max_length=max_length, strip_whitespace=True)`

Usage example:

```
class Foo:
    bar: StrLimited[10] # bar is a str with maximum length of 10 characters
```

## Related issue number

fix #2104

## Checklist

* [x] Unit tests for the changes exist
* [x] Tests pass on CI and coverage remains at 100%
* [x] Documentation reflects the changes where applicable
* [x] `changes/<pull request or issue id>-<github username>.md` file added describing change
  (see [changes/README.md](https://github.com/samuelcolvin/pydantic/blob/master/changes/README.md) for details)
* [x] My PR is ready to review, **please add a comment including the phrase "please review" to assign reviewers**

### CODE REVIEW

**Comment by lucastosetto** [comment]
please review

File: changes/2104-lucastosetto.md [hunk 1]
```diff
@@ -0,0 +1 @@
+Added simplified length limited string type, declared as StrLimited[max_length].
```
Comment (samuelcolvin) [inline changes/2104-lucastosetto.md:1]:
```suggestion
Added simplified length limited string type, declared as `StrLimited[max_length]`.
```

File: pydantic/types.py [hunk 2]
```diff
@@ -436,6 +437,12 @@ def constr(
     return _registered(type('ConstrainedStrValue', (ConstrainedStr,), namespace))
 
 
+class StrLimited:
+    @classmethod
+    def __class_getitem__(cls, max_length: int) -> Type[str]:
```
Comment (samuelcolvin) [inline pydantic/types.py:442]:
```suggestion
    def __class_getitem__(cls, max_length: int) -> Type[ConstrainedStr]:
```

File: pydantic/types.py [hunk 3]
```diff
@@ -436,6 +437,12 @@ def constr(
     return _registered(type('ConstrainedStrValue', (ConstrainedStr,), namespace))
 
 
+class StrLimited:
```
Comment (samuelcolvin) [inline pydantic/types.py:440]: you might want to implement `__get_validators__` and raise an error people try to use `StrLimited` with parameterising it.

File: tests/test_types.py [hunk 4]
```diff
@@ -733,6 +734,34 @@ class Model(BaseModel):
     ]
 
 
+def test_str_limited_good():
+    max_length = 5
+
+    class Model(BaseModel):
+        v: StrLimited[max_length]
```
Comment (samuelcolvin) [inline tests/test_types.py:741]: cleanest to hard code the length
```suggestion
        v: StrLimited[5]
```

File: tests/test_types.py [hunk 5]
```diff
@@ -733,6 +734,34 @@ class Model(BaseModel):
...
+
+    m = Model(v='abcde')
```
Comment (samuelcolvin) [inline tests/test_types.py:743]:
```suggestion
    m = Model(v='abcde ')
```
show that it also strips white space.

File: tests/test_types.py [hunk 6]
```diff
@@ -733,6 +734,34 @@ class Model(BaseModel):
...
+    assert m.v == 'abcde'
+
+
+def test_str_limited_too_long():
+    max_length = 5
+
+    class Model(BaseModel):
+        v: StrLimited[max_length]
```
Comment (samuelcolvin) [inline tests/test_types.py:751]:
```suggestion
        v: StrLimited[5]
```
same, always best to hard code tests unless you need to vary the value, which you don't here.

File: pydantic/types.py (hunk 2 above)
Comment (samuelcolvin) [inline pydantic/types.py:442]: We could also be clever and allow slice usage:

```py
class MyModel(BaseModel):
    has_max_length: StrLimited[10]
    has_min_and_max_length: StrLimited[5:10]
```

WDYT?

**Review by samuelcolvin — COMMENTED** [review]
Looking good, will need to be added to docs.

**Comment by samuelcolvin** [comment]
please update.

File: tests/test_types.py [hunk 7]
```diff
@@ -733,6 +734,34 @@ class Model(BaseModel):
     ]
 
 
+def test_str_limited_good():
+    max_length = 5
```
Comment (samuelcolvin) [inline tests/test_types.py:738]: you can remove this, it's not used.

File: tests/test_types.py [hunk 8]
```diff
@@ -733,6 +734,34 @@ class Model(BaseModel):
...
+
+    class Model(BaseModel):
+        v: StrLimited[5]
+
+    m = Model(v='abcde ')
+    assert m.v == 'abcde'
+
+
+def test_str_limited_too_long():
+    max_length = 5
```
Comment (samuelcolvin) [inline tests/test_types.py:748]: you can remove this and just use a hard-coded value below, it's not adding anything.

**Review by samuelcolvin — CHANGES_REQUESTED** [review]
Looking good, but a few other things still to do:
* `__get_validators__` as suggested above
* add docs, as suggested above

**Comment by samuelcolvin** [comment]
@lucastosetto any chance you can fix this?

**Comment by samuelcolvin** [comment]
See [twitter :bird:](https://twitter.com/pydantic/status/1560554679473303552), I've you'd like this to be included in V1.10, please fix it and request a review **TODAY**.

Or if you need this in V1.10 but don't have time to complete it (or aren't the author), please comment here and on #4324

**Comment by cloudwalk-tests** [comment]
> @lucastosetto any chance you can fix this?

@samuelcolvin Sure, I can work on it this week, but I won't have time to do this for v1.10

**Comment by lucastosetto** [comment]
> @lucastosetto any chance you can fix this?

Sure, I can work on it this week, but I won't have time to do it for v1.10 release.

**Comment by samuelcolvin** [comment]
Ok, I might try to finish it for you.

If it doesn't make it into v1.10 much of the implementation will have to change for V2.

**Comment by samuelcolvin** [comment]
> Ok, I might try to finish it for you.

Ignore that, I'm not going to have time to work on this, I think it'll have to wait for V2.

File: pydantic/types.py [hunk 9]
```diff
@@ -436,6 +437,12 @@ def constr(
...
+    @classmethod
+    def __class_getitem__(cls, max_length: int) -> Type[ConstrainedStr]:
+        return constr(max_length=max_length, strip_whitespace=True)
```
Comment (antonagestam) [inline pydantic/types.py:443]: This usage of `__class_getitem__` isn't compatible with typing, at least mypy will give an error for `StrLimited[10]`, and I believe the other major type checkers would too. Mypy gives: `error: The type "Type[StrLimited]" is not generic and not indexable [misc]`. I'd suggest redesigning this to make `StrLimited` an object that implements `__get_item__` instead.

**Pull request closed without merging**